src/olcs/overlay.js

```diff
@@ -0,0 +1,350 @@
+    const sourceProjection = this.parent_.getMap().getView().getProjection();
+    const coords = ol.proj.transform(position, sourceProjection, 'EPSG:4326');
+    this.setPosition(coords);
+    this.handleMapChanged();
+  } else {
+    this.setPosition(undefined);
+    this.setVisible(false);
+  }
+};
+
+/**
+ * @private
+ */
+olcs.Overlay.prototype.changeElement_ = function() {
+  function cloneNode(node, parent) {
+    const clone = node.cloneNode();
+    if (parent) {
+      parent.appendChild(clone);
+    }
+    if (node.nodeType != Node.TEXT_NODE) {
+      clone.addEventListener('click', (event) => {
+        node.dispatchEvent(new MouseEvent('click', event));
+        event.stopPropagation();
+      });
+    }
+    const nodes = node.childNodes;
+    for (let i = 0; i < nodes.length; i++) {
+      if (!nodes[i]) {
+        continue;
+      }
+      cloneNode(nodes[i], clone);
+    }
+    return clone;
+  }
+
+  if (this.parent_.getElement()) {
+    const clonedNode = cloneNode(this.parent_.getElement(), null);
+
+    this.setElement(clonedNode);
+    const parentNode = this.getElement().parentNode;
+    if (parentNode) {
+      this.addMutationObserver_(parentNode);
+      while (parentNode.firstChild) {
+        parentNode.removeChild(parentNode.firstChild);
+      }
+      const childNodes = this.parent_.getElement().parentNode.childNodes;
+      for (let i = 0; i < childNodes.length; i++) {
+        cloneNode(childNodes[i], parentNode);
+      }
+    }
+  }
+};
+
+/**
+ * Update pixel position.
+ * @suppress {checkTypes, accessControls}
+ */
+olcs.Overlay.prototype.updatePixelPosition = function() {
+  const scene = this.getScene();
+  const position = this.getPosition();
+  if (!scene || !position) {
+    this.setVisible(false);
+    return;
+  }
+  let cartesian;
+  if (position.length === 2) {
+    cartesian = Cesium.Cartesian3.fromDegreesArray(position)[0];
+  } else {
+    cartesian = Cesium.Cartesian3.fromDegreesArrayHeights(position)[0];
+  }
+
+  const ellipsoidBoundingSphere = new Cesium.BoundingSphere(new Cesium.Cartesian3(), 6356752);
+  const occluder = new Cesium.Occluder(ellipsoidBoundingSphere, scene.camera.position);
+  // check if overlay position is behind the horizon
+  if (!occluder.isPointVisible(cartesian)) {
+    this.setVisible(false);
+    return;
+  }
+  const cullingVolume = scene.camera.frustum.computeCullingVolume(scene.camera.position, scene.camera.direction, scene.camera.up);
+  // check if overlay position is visible from the camera
+  if (cullingVolume.computeVisibility(new Cesium.BoundingSphere(cartesian)) !== 1) {
+    this.setVisible(false);
+    return;
+  }
+  this.setVisible(true);
+
+  const pixelCartesian = scene.cartesianToCanvasCoordinates(cartesian);
+  const pixel = [pixelCartesian.x, pixelCartesian.y];
+  const mapSize = [scene.canvas.width, scene.canvas.height];
+  const that = /** @type {ol.Overlay} */ (this);
+  ol.Overlay.prototype.updateRenderedPosition.call(that, pixel, mapSize);
```

Comment:
Yeah we also thought about this, but for this to work we have to either change the Overlay property visibility to "protected" from "private", or we can disable the closure compiler "visibility" Errors like this in the olcesium.json. This is also why we didn't do some of the todos in the code.

"jscomp_off": [
      "lintChecks",
      "analyzerChecks",
      "visibility"
    ],

So, what is your preference, should i disable the compiler flag, or should i create a openlayers PR which changes all ol.Overlay "@private" to "@protected". Then i can refactor the olcs.overlay to directly inherit from ol.Overlay.

Reply:
You should send a PR to OpenLayers to change the visibility to `@protected`. It is something we do when it appears there is a real use-case for it.

Reply:
Hmm, i just checked this, changing properties to @protected is unfortunately not enough because i need to access all the properties of the original Overlay in the new olcs.Overlay. To do this we have three options
- change propertie visibility ( insertFirst_, stopEvent_, ect.) to @public
- add getters properties which do not have them yet 
- add getOptions function which returns an options Object (my preferred  solution)

Which option do you prefer?